File: packages/javascript/lib/format/numberFormat.ts
const DIGITS_ONLY = /^\d+$/;

// Creating an `Intl.NumberFormat` is expensive, so build it once and reuse it.
const FALLBACK_FORMATTER = new Intl.NumberFormat('en-US', {
	roundingPriority: 'morePrecision'
});

// Group digits straight from the string so values beyond `Number.MAX_SAFE_INTEGER`
// keep every digit (`parseInt` would round them).
function groupThousands(digits: string): string {
	let result = '';

	for (let end = digits.length; end > 0; end -= 3) {
		const chunk = digits.slice(Math.max(0, end - 3), end);

		result = result === '' ? chunk : `${chunk},${result}`;
	}

	return result;
}

export function numberFormat(number: number | string): string {
	if (number === null || number === undefined) {
		return '';
	}

	const str = typeof number === 'string' ? number : number.toString();
	const isNegative = str.startsWith('-');
	const abs = isNegative ? str.slice(1) : str;

	const numberParts = abs.split('.');

	const numberFormatted = DIGITS_ONLY.test(numberParts[0])
		? groupThousands(numberParts[0])
		: FALLBACK_FORMATTER.format(parseInt(numberParts[0], 10));

	const result = `${numberFormatted}${numberParts.length > 1 ? `.${numberParts[1]}` : ''}`;
	return isNegative ? `-${result}` : result;
}

File: packages/python/qsu/format/numberFormat.py
from typing import Union


def _numStr(value: Union[int, float]) -> str:
	if isinstance(value, float) and value.is_integer():
		return str(int(value))
	return str(value)


def numberFormat(number: Union[int, float, str, None]) -> str:
	if number is None:
		return ''

	str_value = number if isinstance(number, str) else _numStr(number)
	isNegative = str_value.startswith('-')
	abs_value = str_value[1:] if isNegative else str_value

	numberParts = abs_value.split('.')

	numberFormatted = f'{int(numberParts[0], 10):,}'

	result = f"{numberFormatted}{'.' + numberParts[1] if len(numberParts) > 1 else ''}"
	return f'-{result}' if isNegative else result
